Mono.Debugging/Mono.Debugging.Evaluation/AsyncOperationManager.cs

```diff
@@ -27,215 +27,163 @@
+				task = mc.InvokeAsync ();
+				currentOperations.Add (mc);
 			}
 
-			if (timeout > 0) {
-				if (!methodCall.WaitForCompleted (timeout)) {
-					bool wasAborted = methodCall.Aborted;
-					methodCall.InternalAbort ();
-					lock (operationsToCancel) {
-						operationsToCancel.Remove (methodCall);
-						ST.Monitor.PulseAll (operationsToCancel);
+			bool cancelledAfterTimeout = false;
+			try {
+				if (task.Wait (timeout)) {
+					DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} succeeded in {1} ms", description, timeout));
+					return task.Result;
+				}
+				DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} timed out after {1} ms. Cancelling.", description, timeout));
+				mc.Abort ();
+				try {
+					WaitAfterCancel (mc);
+				}
+				catch (Exception e) {
+					if (IsOperationCancelledException (e)) {
+						DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} was cancelled after timeout", description));
+						cancelledAfterTimeout = true;
 					}
-					if (wasAborted)
-						throw new EvaluatorAbortedException ();
-					else
-						throw new TimeOutException ();
+					throw;
 				}
+				DebuggerLoggingService.LogMessage (string.Format ("{0} cancelling timed out", description));
+				throw new TimeOutException ();
 			}
-			else {
-				methodCall.WaitForCompleted (System.Threading.Timeout.Infinite);
-			}
-
-			lock (operationsToCancel) {
-				operationsToCancel.Remove (methodCall);
-				ST.Monitor.PulseAll (operationsToCancel);
-				if (methodCall.Aborted) {
+			catch (Exception e) {
+				if (IsOperationCancelledException (e)) {
+					if (cancelledAfterTimeout)
+						throw new TimeOutException ();
+					DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} was cancelled outside before timeout", description));
 					throw new EvaluatorAbortedException ();
 				}
+				throw;
+			}
+			finally {
+				lock (currentOperations) {
+					currentOperations.Remove (mc);
+				}
 			}
+		}
+
 
-			if (!string.IsNullOrEmpty (methodCall.ExceptionMessage)) {
-				throw new Exception (methodCall.ExceptionMessage);
+		public event EventHandler<BusyStateEventArgs> BusyStateChanged = delegate {  };
+
+		void ChangeBusyState (bool busy, string description)
+		{
+			try {
+				BusyStateChanged (this, new BusyStateEventArgs {IsBusy = busy, Description = description});
+			}
+			catch (Exception e) {
+				DebuggerLoggingService.LogError ("Exception during ChangeBusyState", e);
 			}
 		}
-		
-		public void Dispose ()
+
+		void WaitAfterCancel (IAsyncOperationBase op)
 		{
-			Disposing = true;
-			lock (operationsToCancel) {
-				foreach (AsyncOperation op in operationsToCancel) {
-					op.InternalShutdown ();
+			var desc = op.Description;
+			DebuggerLoggingService.LogMessage (string.Format ("Waiting for cancel of invoke {0}", desc));
+			if (!op.RawTask.Wait (ShortCancelTimeout)) {
+				try {
+					ChangeBusyState (true, desc);
+					while (true) {
```

Comment:
>Logic for switching to busy state is now different. Please use old logic.

Can you please explain? 
For me the current logic is quite the same:
1. StartInvoke
2. Wait normally for given timeout
3. If timed out -> Abort()
4.  Wait for Short timeout
5. If timed out -> Enter busy
6. Repeat from 4
7. If awaited -> Exit busy